Add bookings

EvgeniyaKuzmina opened the pull request.

EvgeniyaKuzmina committed:
feat: Подключены БД, JPA и Hibernate
Созданы таблицы БД
Реализованы новые репозитории для  сущностей User, Item, Booking.Реализация старых репозиториев была удалена
Переписаны классы сервисы согласно методам новых репозиториев.
Добавлена сущность Comments
Добавлены методы работы с бронированиям, отвечающие за:
- добавление нового запроса на бронирование
- подтверждение или отклонение бронирования владельцем
- получение данных о конкретном бронировании
- получение списка всех бронирований текущего пользователя с фильтром по статусу
- получение списка всех бронирований конкретного пользователя
- получение бронирования по Id

EvgeniyaKuzmina committed:
feat: доработан функционал по добавлению комментария для вещи. Полностью реализована логика для создания и управления бронированием.
style: изменена структура проекта. Созданы дополнительный класс и интерфейс для комментариев.
Выполнение кода отвечает всем требованиям тестов

EvgeniyaKuzmina committed:
feat: подгоняю под тесты

EvgeniyaKuzmina committed:
feat: подгоняю под тесты

EvgeniyaKuzmina committed:
feat: подгоняю под тесты

EvgeniyaKuzmina committed:
feat: подгоняю под тесты

EvgeniyaKuzmina committed:
feat: подгоняю под тесты

EvgeniyaKuzmina committed:
feat: подгоняю под тесты

EvgeniyaKuzmina committed:
feat: подгоняю под тесты

EvgeniyaKuzmina committed:
feat: подгоняю под тесты

EvgeniyaKuzmina commented:
Добрый день.
Хотелось бы заранее объясниться. Нам  сказали что пока тесты можно не учитывать при сдаче работы. Но я все же проверила свою работы согласно тестам отсюда https://github.com/yandex-praktikum/java-shareit/blob/add-bookings/postman/sprint.json
Локально тесты все прошли, скриншот по ссылке http://joxi.ru/8AnRo9jUNQ7WlA 
А при запуске ПР два теста падают. Я не понимаю почему, судя по тесту тестам, я получаю ровно то что ждёт от меня программа. Поэтому прошу понять и простить.

EvgeniyaKuzmina committed:
feat: обновила файлы

Troitsa commented on src/main/resources/application.properties:
```diff
@@ -5,7 +5,10 @@ spring.jpa.hibernate.show_sql=true
 
 # TODO Append connection to DB
-
+spring.datasource.driverClassName=org.h2.Driver
```
В этом спринте подключаем базу данных Postgresql,соответственно драйвер должен быть для Постгреса и датасорс тоже меняется:

spring.datasource.driverClassName=org.postgresql.Driver
spring.datasource.url=jdbc:postgresql://localhost:5432/shareit

EvgeniyaKuzmina replied:
исправила

Troitsa commented on src/main/java/ru/practicum/shareit/user/UserServiceImpl.java:
```diff
@@ -0,0 +1,62 @@
+    @Override
+    public User createUser(UserDto userDto) {
+        User user = UserMapper.toUser(userDto);
+        return userRepository.save(user);
```
можно попробовать так обработать ошибки повторного добавления юзеров с одинаковым емейлом:
(только ConflictException отнаследовать от RuntimeException)
try {
            log.info("Добавлен пользователь {}.", user);
            return userRepository.save(user);
        } catch (DataIntegrityViolationException e) {
            log.error("Пользователь с таким email {} уже существует.", user.getEmail());
            throw new ConflictException(String.format("Пользователь с таким email %s уже существует.",
                    user.getEmail()));
        }

EvgeniyaKuzmina replied:
добавила проверку в оба метода (в создание и в обновление пользователя)

Troitsa commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,21 +1,55 @@
+    @OneToOne
+    @JoinColumn(name = "owner_id", nullable = false)
     private User owner;
+    @Column(name = "request_id")
```
в базе данных у Вас request_id bigint, а здесь String..

EvgeniyaKuzmina replied:
исправила

EvgeniyaKuzmina committed:
feat: 1. Добавлена обработка ошибки, если попытаться создать или обновить пользователя, указав email, который уже принадлежит другому пользователю.
2. Изменён тип данных в объекте Item для request.
3. Подключена БД PostgreSQL вместо H2
4. Изменена сортировка для получения списка всех вещей пользователя

EvgeniyaKuzmina committed:
feat: удалён лишний файл
